### DISCUSSION TITLE
[Ruby][CHI] send data two times for one single Read request in memory controller

### DISCUSSION THREAD

**Discussion by gxflying** [gem5/gem5#595] [Q&A]
Hi sir:
    in the implementation of CHI-mem.sm , I found that for a **ReadNoSnp** to ruby memory controller the data returned fron memory will be send two times when the controller state goes : 
EVENT: --------ReadNoSnp--------------MemoryData-----------------------Trigger_Send-----------------------Trigger_SendDone
STATE : READY -----------> READ_MEM ------------> SENDING_NET_DATA -------------> SENDING_NET_DATA ------------------> READY
In both transitions triggered  by **MemoryData** and **Trigger_Send** will ivoke **sendDataAndCheck** which will enqueue message into *m_datOut_ptr,  and as a result , the msg will be send out from the datOut port twice ?!

> ((*m_datOut_ptr)).enqueue(out_msg, clockEdge(), cyclesToTicks(Cycles(m_data_latency)));

```
  transition(READING_MEM, MemoryData, SENDING_NET_DATA) {
    prepareSend;
    sendDataAndCheck;
    popMemoryQueue;
  }
  transition(SENDING_NET_DATA, Trigger_Send) {
    sendDataAndCheck;
    popTriggerQueue;
  }
```

https://github.com/gem5/gem5/blob/d0d3c74ce00f6d852a39060f3b7ce083887b1cfd/src/mem/ruby/protocol/chi/CHI-mem.sm#L727
https://github.com/gem5/gem5/blob/d0d3c74ce00f6d852a39060f3b7ce083887b1cfd/src/mem/ruby/protocol/chi/CHI-mem.sm#L733

please help to figure out whether this behavior is expected ?  thanks~

**Comment by powerjg** [comment]
@tiagormk do you know the answer to this question?

**Comment by tiagormk** [comment]
Yes, that's expected.
The number of messages sent depends on the amount of bytes being read an the size of the CHI data message. I believe the default size of the CHI data message is 32 bytes, and usually each request will read one cache line (64 bytes), so CHI-mem will split the data into two messages.

**Comment by gxflying** [comment]
Hi @tiagormk thanks for your clarificaiton
and I find the default size of mem channel size is indeed set to 32
what is the size is NOT 32 bytes?  do it means we have to modified the source code in the aforementioned transition ？
for example 16 bytes channel size , we must modify the source code and call "**sendDataAndCheck**"  4 times (for a full cache line )?
https://github.com/gem5/gem5/blob/461af5157552d916baec732dfc7d8e97a2e781d8/src/python/gem5/components/cachehierarchies/chi/nodes/memory_controller.py#L69

**Reply by tiagormk** [reply]
it shouldn't require source code changes. sendDataAndCheck will be automatically triggered multiple times to send all necessary messages (see https://github.com/gem5/gem5/blob/d0d3c74ce00f6d852a39060f3b7ce083887b1cfd/src/mem/ruby/protocol/chi/CHI-mem.sm#L626).

**Reply by gxflying** [reply]
yeah, it seems that the  sendDataAndCheck will handle multi beat data request, 
then my question is what I asked the very fist,
why  the flow will **explicitly** call the sendDataAndCheck  **twice**？  while sendDataAndCheck can handle this automatically 

1. one when transitoon from READING_MEM to SENDING_NET_DATA
2. one when SENDING_NET_DATA and triggered by Trigger_Send

```
  transition(READING_MEM, MemoryData, SENDING_NET_DATA) {
    prepareSend;
    sendDataAndCheck;
    popMemoryQueue;
  }
  transition(SENDING_NET_DATA, Trigger_Send) {
    sendDataAndCheck;
    popTriggerQueue;
  }
```

**Comment by gxflying** [comment]
Oh I see, the second sendDataAndCheck  in my quesition is triggered by the sendDataAndCheck !
thanks I will close this issue~
